packages/user/src/User/Login.purs

```diff
@@ -183,26 +192,49 @@ failOnEmailMismatch self email
       liftEffect $ self.setState _ { errors { login = Just LoginEmailMismatchError } }
       throwError $ error "Emails don't match"
 
-onLogin :: Self -> Effect Unit
-onLogin self@{ props, state } = props.launchAff_ do
-  user <- User.loginTraditional
-            { username: fromMaybe "" state.formEmail
-            , password: fromMaybe "" state.formPassword
+onLogin :: Self -> Form.ValidatedForm LoginField LoginForm -> Effect Unit
+onLogin self@{ props, state } = unV
+  (\errors -> do
+      self.setState _
+        { formEmail    = state.formEmail    <|> Just ""
+        , formPassword = state.formPassword <|> Just ""
+        })
+  logUserIn
+  where
+    logUserIn validForm
+      | Just validUsername <- validForm.username
+      , Just validPassword <- validForm.password
+      , not String.null $ validUsername
+      , not String.null $ validPassword
+      = props.launchAff_ do
+        user <- User.loginTraditional
+            { username: validUsername
+            , password: validPassword
             , mergeToken: toNullable $ map _.token state.merge
             }
-  case user of
-    Right _ ->
-      -- removing merge token from state in case of success
-      liftEffect $ self.setState _ { merge = Nothing }
-    Left LoginInvalidCredentials ->
-      liftEffect $ self.setState _ { errors { login = Just LoginInvalidCredentials } }
-    Left SomethingWentWrong ->
-      liftEffect $ self.setState _ { errors { login = Just SomethingWentWrong } }
-    Left _ ->
-      throwError $ error "An unexpected error occurred during traditional login"
-  -- This call needs to be last, as it will unmount the login component.
-  -- (We cannot set state of an unmounted component)
-  liftEffect $ props.onUserFetch user
+        case user of
+          Right _ ->
+            -- removing merge token from state in case of success
+            liftEffect $ self.setState _ { merge = Nothing }
+          Left LoginInvalidCredentials ->
+            liftEffect $ self.setState _ { errors { login = Just LoginInvalidCredentials } }
+          Left SomethingWentWrong ->
+            liftEffect $ self.setState _ { errors { login = Just SomethingWentWrong } }
+          Left _ ->
+            throwError $ error "An unexpected error occurred during traditional login"
+        -- This call needs to be last, as it will unmount the login component.
+        -- (We cannot set state of an unmount
+        liftEffect $ props.onUserFetch user
```

Comment:
Should we do it in case if the `Left _` branch matches and exception is thrown?

Reply:
Yes 👽